### PR TITLE
executor: support Chunk in LimitExec

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: util/chunk/chunk.go [hunk 1]
```diff
@@ -145,6 +148,54 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
...
 }
 
+// Copy copys rows ranged ["begin", "end") in "other" to current Chunk.
+func (c *Chunk) Copy(other *Chunk, begin, end int) {
+	for colID, src := range other.columns {
+		dst := c.columns[colID]
+		if src.isFixed() {
+			elemLen := len(src.elemBuf)
+			if len(dst.data) < elemLen*(end-begin) {
+				dst.data = make([]byte, elemLen*(end-begin), elemLen*(end-begin))
```
Comment (coocood) [inline util/chunk/chunk.go]: If we make data here, we would be able to reuse the memory.

File: util/chunk/chunk.go [hunk 2]
```diff
@@ -145,6 +148,54 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
...
+			}
+			copy(dst.ifaces, src.ifaces[begin:end])
+			dst.ifaces = dst.ifaces[:end-begin]
+		}
+		dst.length, dst.nullCount = 0, 0
+		numBytesInBitmap := (end - begin + 7) >> 3
+		if len(dst.nullBitmap) < numBytesInBitmap {
+			dst.nullBitmap = make([]byte, numBytesInBitmap)
+		}
+		for i := begin; i < end; i++ {
```
Comment (coocood) [inline util/chunk/chunk.go]: We should be able to append the bitmap bytes directly, in the case of not aligned, we can shift the bitmap.

File: util/chunk/chunk.go [hunk 3]
```diff
@@ -145,6 +148,51 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
...
+
+// Truncate truncates "tailRows" rows from the tail..
+func (c *Chunk) Truncate(tailRows int) {
+	for _, col := range c.columns {
+		if col.isFixed() {
+			elemLen := len(col.elemBuf)
+			col.data = col.data[:(col.length-tailRows)*elemLen]
+		} else if col.isVarlen() {
+			col.data = col.data[:col.offsets[col.length-tailRows+1]]
+			col.offsets = col.offsets[:col.length-tailRows]
```
Comment (coocood) [inline util/chunk/chunk.go]: The length of `offsets` is larger than `col.length` by 1.

**Comment by coocood** [comment]
@zz-jason 
Please add tests for `Append` and `Truncate`.

File: util/chunk/chunk_test.go [hunk 4]
```diff
@@ -122,6 +123,127 @@ func (s *testChunkSuite) TestChunk(c *C) {
...
+
+	src := NewChunk(fieldTypes)
+	src.AppendFloat32(0, 12.8)
+	src.AppendString(1, "abc")
+	src.AppendJSON(2, jsonObj)
+	src.AppendNull(0)
+	src.AppendNull(1)
+	src.AppendNull(2)
+
+	for i := 0; i < 7; i++ {
```
Comment (coocood) [inline util/chunk/chunk_test.go]: Why not `i < 8` to save the above 6 lines?

File: executor/executor.go [hunk 5]
```diff
@@ -370,14 +372,61 @@ func (e *LimitExec) Next(goCtx goctx.Context) (Row, error) {
...
+		if batchSize == 0 {
+			return nil
+		}
+		e.cursor += batchSize
+		if e.cursor >= e.begin {
+			e.meetFirstBatch = true
+			if e.cursor > e.end {
+				batchSize -= e.cursor - e.end
+			}
+			fmt.Printf("matched, batchSize: %v\n", batchSize)
```
Comment (coocood) [inline executor/executor.go]: remove the debug log.

File: executor/executor.go [hunk 6]
```diff
@@ -370,14 +372,60 @@ func (e *LimitExec) Next(goCtx goctx.Context) (Row, error) {
...
+		err := e.children[0].NextChunk(e.childrenResults[0])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		batchSize := uint64(e.childrenResults[0].NumRows())
+		// no more data.
+		if batchSize == 0 {
+			return nil
+		}
+		e.cursor += batchSize
```
Comment (coocood) [inline executor/executor.go]: I think to update the cursor after we append the data is easier to understand.
Then the cursor will never be greater than e.end.

Comment (zz-jason) [reply executor/executor.go]: It's not that hard to understand and if we update cursor after appending data the code will be more or less ugly.

Comment (coocood) [reply executor/executor.go]: I think there is a way to make the code easier to understand.

File: executor/executor.go [hunk 7]
```diff
@@ -370,14 +372,63 @@ func (e *LimitExec) Next(goCtx goctx.Context) (Row, error) {
...
+
 // Open implements the Executor Open interface.
 func (e *LimitExec) Open(goCtx goctx.Context) error {
-	e.Idx = 0
-	return errors.Trace(e.children[0].Open(goCtx))
+	if err := e.baseExecutor.Open(goCtx); err != nil {
+		return errors.Trace(err)
+	}
+	e.cursor = 0
+	e.meetFirstBatch = false
```
Comment (coocood) [inline executor/executor.go]: If the offset is 0, we can set meetFirstBatch to true

File: util/chunk/chunk.go [hunk 8]
```diff
@@ -146,6 +149,51 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
...
+		}
+		for i := begin; i < end; i++ {
+			dst.appendNullBitmap(!src.isNull(i))
+			dst.length++
+		}
+	}
+}
+
+// Truncate truncates "tailRows" rows from the tail..
+func (c *Chunk) Truncate(tailRows int) {
```
Comment (coocood) [inline util/chunk/chunk.go]: I think truncate to length is more commonly used.

File: executor/executor.go [hunk 9]
```diff
@@ -370,14 +372,63 @@ func (e *LimitExec) Next(goCtx goctx.Context) (Row, error) {
...
+	err := e.children[0].NextChunk(chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	batchSize := uint64(chk.NumRows())
+	// no more data.
+	if batchSize == 0 {
+		return nil
+	}
+	if e.cursor > e.end {
```
Comment (coocood) [inline executor/executor.go]: Should update cursor before this check.
Seems not covered by tests.

**Comment by coocood** [comment]
@zz-jason 
We need to write test cases for multiple chunk limit.
Our existing tests usually have a small row count, less than one chunk.

**Comment by zz-jason** [comment]
@coocood we can add a session variable to control max chunk capacity, before testing, we can set a small chunk size, like 2.

**Comment by coocood** [comment]
LGTM

File: util/chunk/chunk.go [hunk 10]
```diff
@@ -146,6 +149,51 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
 	}
 }
 
+// Append appends rows in [begin, end) in another Chunk to a Chunk.
+func (c *Chunk) Append(other *Chunk, begin, end int) {
```
Comment (XuHuaiyu) [inline util/chunk/chunk.go]: should we check the validation of begin and end first.

File: util/chunk/chunk.go [hunk 11]
```diff
@@ -146,6 +149,51 @@ func (c *Chunk) AppendRow(colIdx int, row Row) {
...
+		}
+		for i := begin; i < end; i++ {
+			dst.appendNullBitmap(!src.isNull(i))
+			dst.length++
+		}
+	}
+}
+
+// TruncateTo truncates rows from tail to head in a Chunk to "numRows" rows.
+func (c *Chunk) TruncateTo(numRows int) {
```
Comment (XuHuaiyu) [inline util/chunk/chunk.go]: check validation of numRows?

File: util/chunk/chunk.go [hunk 12]
```diff
@@ -51,23 +51,26 @@ func NewChunk(fields []*types.FieldType) *Chunk {
...
 	c.columns = append(c.columns, &column{
-		offsets: []int32{0},
-		data:    make([]byte, 0, initCap),
+		offsets:    make([]int32, 1, initCap+1),
+		data:       make([]byte, 0, initCap*4),
+		nullBitmap: make([]byte, 0, initCap>>3),
 	})
 }
 
 // addInterfaceColumn adds an interface column which holds element as interface.
```
Comment (XuHuaiyu) [inline util/chunk/chunk.go]: update the column, not `an interface column`
so as the other comments.

File: util/chunk/chunk.go (hunk 10 above)
Comment (zz-jason) [reply util/chunk/chunk.go]: no need to, all Chunk's API should not do the validation work.

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**